ports-of-call/static_vector.hpp

```diff
@@ -0,0 +1,436 @@
+    using pointer = value_type *;
+    using reference = value_type &;
+
+    using const_iterator_type = iterator_type<Element_ const>;
+
+    PORTABLE_FUNCTION constexpr iterator_type() : it_{nullptr} {}
+
+    PORTABLE_FUNCTION constexpr iterator_type(Element_ *it) : it_{it} {}
+
+    // const iterator from non-const iterator
+    template <typename U = value_type,
+              std::enable_if_t<std::is_const<U>::value> * = nullptr>
+    PORTABLE_FUNCTION constexpr iterator_type(iterator_type<std::remove_const_t<U>> it)
+        : it_{it.it_} {}
+
+   private:
+    Element_ *it_;
+
+    // non-const iterator from const iterator...generally dangerous
+    // (used for convinience in static_vector implementation)
+    template <typename U = value_type,
+              std::enable_if_t<not std::is_const<U>::value> * = nullptr>
+    PORTABLE_FUNCTION constexpr iterator_type(iterator_type<std::add_const_t<U>> it)
+        : it_{const_cast<U *>(it.it_)} {}
+
+   public:
+    PORTABLE_FUNCTION constexpr iterator_type &operator++() {
+      ++(this->it_);
+      return *this;
+    }
+
+    PORTABLE_FUNCTION constexpr iterator_type operator++(int) const {
+      iterator_type temp{*this};
+      ++temp;
+      return temp;
+    }
+
+    PORTABLE_FUNCTION constexpr iterator_type &operator--() {
+      --(this->it_);
+      return *this;
+    }
+
+    PORTABLE_FUNCTION constexpr iterator_type operator--(int) const {
+      iterator_type temp{*this};
+      --temp;
+      return temp;
+    }
+
+    PORTABLE_FUNCTION constexpr iterator_type &operator+=(difference_type m) {
+      this->it_ += m;
+      return *this;
+    }
+
+    PORTABLE_FUNCTION constexpr iterator_type &operator-=(difference_type m) {
+      return *this += -m;
+    }
+
+    PORTABLE_FUNCTION constexpr iterator_type operator+(difference_type m) const {
+      iterator_type temp{*this};
+      return temp += m;
+    }
+
+    PORTABLE_FUNCTION constexpr iterator_type operator-(difference_type m) const {
+      iterator_type temp{*this};
+      return temp -= m;
+    }
+
+    PORTABLE_FUNCTION constexpr difference_type
+    operator-(const_iterator_type other) const {
+      return this->it_ - other.it_;
+    }
+
+    PORTABLE_FUNCTION constexpr bool operator==(const_iterator_type other) const {
+      return this->it_ == other.it_;
+    }
+
+    PORTABLE_FUNCTION constexpr bool operator!=(const_iterator_type other) const {
+      return not(this->it_ == other.it_);
+    }
+
+    PORTABLE_FUNCTION constexpr bool operator<(const_iterator_type other) const {
+      return this->it_ < other.it_;
+    }
+
+    PORTABLE_FUNCTION constexpr bool operator>(const_iterator_type other) const {
+      return other < *this;
+    }
+
+    PORTABLE_FUNCTION constexpr bool operator>=(const_iterator_type other) const {
+      return not(*this < other);
+    }
+
+    PORTABLE_FUNCTION constexpr bool operator<=(const_iterator_type other) const {
+      return not(*this > other);
+    }
```

Comment:
Are all of the comparators really necessary? Won't the compiler fill in the definitions of some of these if we don't specify?

Reply:
I think you're thinking of the [spaceship operator](https://en.cppreference.com/w/cpp/language/default_comparisons), which a C++20 feature.